File: library/src/test/java/me/proxer/library/api/DelimitedEnumSetAdapterFactoryTest.java
package me.proxer.library.api;

import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.JsonDataException;
import com.squareup.moshi.Moshi;
import com.squareup.moshi.Types;
import lombok.Value;
import me.proxer.library.entity.info.Entry;
import me.proxer.library.enums.FskConstraint;
import me.proxer.library.enums.Gender;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.lang.reflect.ParameterizedType;
import java.util.Collections;
import java.util.EnumSet;
import java.util.List;
import java.util.Set;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;

/**
 * @author Ruben Gees
 */
class DelimitedEnumSetAdapterFactoryTest {

    private DelimitedEnumSetAdapterFactory factory;
    private Moshi moshi;

    @BeforeEach
    void setUp() {
        factory = new DelimitedEnumSetAdapterFactory();
        moshi = new Moshi.Builder()
                .add(factory)
                .build();
    }

    @Test
    void testCreate() throws NoSuchFieldException {
        final DelimitedEnumSet annotation = GenderTestClass.class.getField("genders")
                .getAnnotation(DelimitedEnumSet.class);

        final ParameterizedType type = Types.newParameterizedType(Set.class, Gender.class);
        final JsonAdapter<?> adapter = factory.create(type, Collections.singleton(annotation), moshi);

        assertThat(adapter).isNotNull();
    }

    @Test
    void testCreateInvalidType() throws NoSuchFieldException {
        final DelimitedEnumSet annotation = GenderTestClass.class.getField("genders")
                .getAnnotation(DelimitedEnumSet.class);

        final ParameterizedType type = Types.newParameterizedType(List.class, FskConstraint.class);
        final JsonAdapter<?> adapter = factory.create(type, Collections.singleton(annotation), moshi);

        assertThat(adapter).isNull();
    }

    @Test
    void testCreateInvalidParameterType() throws NoSuchFieldException {
        final DelimitedEnumSet annotation = GenderTestClass.class.getField("genders")
                .getAnnotation(DelimitedEnumSet.class);

        final ParameterizedType type = Types.newParameterizedType(Set.class, Entry.class);
        final JsonAdapter<?> adapter = factory.create(type, Collections.singleton(annotation), moshi);

        assertThat(adapter).isNull();
    }

    @Test
    void testCreateNoParameterType() throws NoSuchFieldException {
        final DelimitedEnumSet annotation = GenderTestClass.class.getField("genders")
                .getAnnotation(DelimitedEnumSet.class);

        final JsonAdapter<?> adapter = factory.create(Set.class, Collections.singleton(annotation), moshi);

        assertThat(adapter).isNull();
    }

    @Test
    void testCreateNoAnnotation() {
        final ParameterizedType type = Types.newParameterizedType(Set.class, Gender.class);
        final JsonAdapter<?> adapter = factory.create(type, Collections.emptySet(), moshi);

        assertThat(adapter).isNull();
    }

    @Test
    void testFromJsonSingle() throws IOException {
        final GenderTestClass result = moshi.adapter(GenderTestClass.class)
                .fromJson("{\"genders\":\"f\"}");

        assertThat(result).isNotNull();
        assertThat(result.genders).containsExactly(Gender.FEMALE);
    }

    @Test
    void testFromJsonMultiple() throws IOException {
        final GenderTestClass result = moshi.adapter(GenderTestClass.class)
                .fromJson("{\"genders\":\"f m\"}");

        assertThat(result).isNotNull();
        assertThat(result.genders).containsExactly(Gender.MALE, Gender.FEMALE);
    }

    @Test
    void testFromJsonInvalidDelimiter() throws IOException {
        final GenderTestClass result = moshi.adapter(GenderTestClass.class)
                .fromJson("{\"genders\":\"f,m\"}");

        assertThat(result).isNotNull();
        assertThat(result.genders).containsExactly(Gender.UNKNOWN);
    }

    @Test
    void testFromJsonEmpty() throws IOException {
        final GenderTestClass result = moshi.adapter(GenderTestClass.class)
                .fromJson("{\"genders\":\"\"}");

        assertThat(result).isNotNull();
        assertThat(result.genders).isEmpty();
    }

    @Test
    void testFromJsonNull() throws IOException {
        final GenderTestClass result = moshi.adapter(GenderTestClass.class)
                .fromJson("{\"genders\":null}");

        assertThat(result).isNotNull();
        assertThat(result.genders).isEmpty();
    }

    @Test
    void testFromJsonDifferentDelimiter() throws IOException {
        final GenderWithDelimiterTestClass result = moshi.adapter(GenderWithDelimiterTestClass.class)
                .fromJson("{\"genders\":\"f, m\"}");

        assertThat(result).isNotNull();
        assertThat(result.genders).containsExactly(Gender.MALE, Gender.FEMALE);
    }

    @Test
    void testFromJsonFallback() throws IOException {
        final GenderTestClass result = moshi.adapter(GenderTestClass.class)
                .fromJson("{\"genders\":\"f m xyz\"}");

        assertThat(result).isNotNull();
        assertThat(result.genders).containsExactly(Gender.MALE, Gender.FEMALE, Gender.UNKNOWN);
    }

    @SuppressWarnings({"CheckReturnValue", "ResultOfMethodCallIgnored"})
    @Test
    void testFromJsonNoFallback() {
        final JsonAdapter<FskConstrainTestClass> adapter = moshi.adapter(FskConstrainTestClass.class);

        assertThatExceptionOfType(JsonDataException.class)
                .isThrownBy(() -> adapter.fromJson("{\"fskConstraints\":\"fsk0 xyz\"}"))
                .withMessageContaining("Expected one of [fsk0")
                .withMessageContaining("but was xyz at path $.fskConstraints");
    }

    @Test
    void testToJsonSingle() {
        final GenderTestClass testSubject = new GenderTestClass(EnumSet.of(Gender.FEMALE));
        final String result = moshi.adapter(GenderTestClass.class).toJson(testSubject);

        assertThat(result).isEqualTo("{\"genders\":\"f\"}");
    }

    @Test
    void testToJsonMultiple() {
        final EnumSet<Gender> values = EnumSet.of(Gender.FEMALE, Gender.MALE);
        final GenderTestClass testSubject = new GenderTestClass(values);
        final String result = moshi.adapter(GenderTestClass.class).toJson(testSubject);

        assertThat(result).isEqualTo("{\"genders\":\"m f\"}");
    }

    @Test
    void testToJsonMultipleDifferentDelimiter() {
        final EnumSet<Gender> values = EnumSet.of(Gender.FEMALE, Gender.MALE);
        final GenderWithDelimiterTestClass testSubject = new GenderWithDelimiterTestClass(values);
        final String result = moshi.adapter(GenderWithDelimiterTestClass.class).toJson(testSubject);

        assertThat(result).isEqualTo("{\"genders\":\"m, f\"}");
    }

    @Value
    private static class GenderTestClass {

        @DelimitedEnumSet
        public Set<Gender> genders;
    }

    @Value
    private static class GenderWithDelimiterTestClass {

        @DelimitedEnumSet(delimiter = ", ")
        public Set<Gender> genders;
    }

    @Value
    private static class FskConstrainTestClass {

        @DelimitedEnumSet
        public Set<FskConstraint> fskConstraints;
    }
}

File: library/src/test/kotlin/me/proxer/library/internal/adapter/DelimitedEnumSetAdapterFactoryTest.kt
package me.proxer.library.internal.adapter

import com.squareup.moshi.JsonClass
import com.squareup.moshi.JsonDataException
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import me.proxer.library.entity.info.Entry
import me.proxer.library.enums.FskConstraint
import me.proxer.library.enums.Gender
import org.amshove.kluent.invoking
import org.amshove.kluent.shouldBeEmpty
import org.amshove.kluent.shouldBeEqualTo
import org.amshove.kluent.shouldBeNull
import org.amshove.kluent.shouldContain
import org.amshove.kluent.shouldNotBeNull
import org.amshove.kluent.shouldThrow
import org.junit.jupiter.api.Test

/**
 * @author Ruben Gees
 */
class DelimitedEnumSetAdapterFactoryTest {

    private val factory = DelimitedEnumSetAdapterFactory()

    private val moshi = Moshi.Builder()
        .add(factory)
        .build()

    private val annotation = GenderTestClass::class.java.getDeclaredField("genders")
        .getAnnotation(DelimitedEnumSet::class.java)

    @Test
    fun testCreate() {
        val type = Types.newParameterizedType(Set::class.java, Gender::class.java)
        val adapter = factory.create(type, setOf(annotation), moshi)

        adapter.shouldNotBeNull()
    }

    @Test
    fun testCreateInvalidType() {
        val type = Types.newParameterizedType(List::class.java, FskConstraint::class.java)
        val adapter = factory.create(type, setOf(annotation), moshi)

        adapter.shouldBeNull()
    }

    @Test
    fun testCreateInvalidParameterType() {
        val type = Types.newParameterizedType(Set::class.java, Entry::class.java)
        val adapter = factory.create(type, setOf(annotation), moshi)

        adapter.shouldBeNull()
    }

    @Test
    fun testCreateNoParameterType() {
        val adapter = factory.create(Set::class.java, setOf(annotation), moshi)

        adapter.shouldBeNull()
    }

    @Test
    fun testCreateNoAnnotation() {
        val type = Types.newParameterizedType(Set::class.java, Gender::class.java)
        val adapter = factory.create(type, emptySet(), moshi)

        adapter.shouldBeNull()
    }

    @Test
    fun testFromJsonSingle() {
        val result = moshi.adapter(GenderTestClass::class.java).fromJson("""{"genders":"f"}""")

        result.shouldNotBeNull()
        result.genders shouldBeEqualTo setOf(Gender.FEMALE)
    }

    @Test
    fun testFromJsonMultiple() {
        val result = moshi.adapter(GenderTestClass::class.java).fromJson("""{"genders":"f m"}""")

        result.shouldNotBeNull()
        result.genders shouldBeEqualTo setOf(Gender.MALE, Gender.FEMALE)
    }

    @Test
    fun testFromJsonInvalidDelimiter() {
        val result = moshi.adapter(GenderTestClass::class.java).fromJson("""{"genders":"f,m"}""")

        result.shouldNotBeNull()
        result.genders shouldBeEqualTo setOf(Gender.UNKNOWN)
    }

    @Test
    fun testFromJsonEmpty() {
        val result = moshi.adapter(GenderTestClass::class.java).fromJson("""{"genders":""}""")

        result.shouldNotBeNull()
        result.genders.shouldNotBeNull()
        result.genders.shouldBeEmpty()
    }

    @Test
    fun testFromJsonNull() {
        val result = moshi.adapter(GenderTestClass::class.java).fromJson("""{"genders":null}""")

        result.shouldNotBeNull()
        result.genders.shouldNotBeNull()
        result.genders.shouldBeEmpty()
    }

    @Test
    fun testFromJsonDifferentDelimiter() {
        val result = moshi.adapter(GenderWithDelimiterTestClass::class.java).fromJson("""{"genders":"f, m"}""")

        result.shouldNotBeNull()
        result.genders shouldBeEqualTo setOf(Gender.MALE, Gender.FEMALE)
    }

    @Test
    fun testFromJsonFallback() {
        val result = moshi.adapter(GenderTestClass::class.java).fromJson("""{"genders":"f m xyz"}""")

        result.shouldNotBeNull()
        result.genders shouldBeEqualTo setOf(Gender.MALE, Gender.FEMALE, Gender.UNKNOWN)
    }

    @Test
    fun testFromJsonNoFallback() {
        val adapter = moshi.adapter(FskConstrainTestClass::class.java)

        val result = invoking { adapter.fromJson("""{"fskConstraints":"fsk0 xyz"}""") } shouldThrow
            JsonDataException::class

        result.exceptionMessage shouldContain "Expected one of [fsk0"
        result.exceptionMessage shouldContain "but was xyz at path $.fskConstraints"
    }

    @Test
    fun testToJsonSingle() {
        val testSubject = GenderTestClass(setOf(Gender.FEMALE))
        val result = moshi.adapter(GenderTestClass::class.java).toJson(testSubject)

        result shouldBeEqualTo """{"genders":"f"}"""
    }

    @Test
    fun testToJsonMultiple() {
        val testSubject = GenderTestClass(setOf(Gender.FEMALE, Gender.MALE))
        val result = moshi.adapter(GenderTestClass::class.java).toJson(testSubject)

        result shouldBeEqualTo """{"genders":"f m"}"""
    }

    @Test
    fun testToJsonMultipleDifferentDelimiter() {
        val testSubject = GenderWithDelimiterTestClass(setOf(Gender.FEMALE, Gender.MALE))
        val result = moshi.adapter(GenderWithDelimiterTestClass::class.java).toJson(testSubject)

        result shouldBeEqualTo """{"genders":"f, m"}"""
    }

    @JsonClass(generateAdapter = true)
    data class GenderTestClass(
        @field:DelimitedEnumSet val genders: Set<Gender>? = null
    )

    @JsonClass(generateAdapter = true)
    data class GenderWithDelimiterTestClass(
        @field:DelimitedEnumSet(delimiter = ", ") val genders: Set<Gender>? = null
    )

    @JsonClass(generateAdapter = true)
    data class FskConstrainTestClass(
        @field:DelimitedEnumSet val fskConstraints: Set<FskConstraint>? = null
    )
}
